runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -921,20 +925,494 @@ static void VMwrtbarEvaluator(TR::Node *node, TR::Register *srcReg, TR::Register
+         }
+      }
+      
+   // Loaded Snippet
+   if (!isInstanceField)
+      {
+      static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->loadSnippet();
+      }
+   else
+      {
+      static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->loadSnippet();
+      }
+   }
+
+TR::Snippet * 
+J9::Power::TreeEvaluator::getFieldWatchInstanceSnippet(TR::CodeGenerator *cg, TR::Node *node, J9Method *m, UDATA loc, UDATA os)
+   {
+   return new (cg->trHeapMemory()) TR::J9PPCWatchedInstanceFieldSnippet(cg, node, m, loc, os);
+   }
+
+TR::Snippet * 
+J9::Power::TreeEvaluator::getFieldWatchStaticSnippet(TR::CodeGenerator *cg, TR::Node *node, J9Method *m, UDATA loc, void *fieldAddress, J9Class *fieldClass)
+   {
+   return new (cg->trHeapMemory()) TR::J9PPCWatchedStaticFieldSnippet(cg, node, m, loc, fieldAddress, fieldClass);
    }
 
 void
-J9::Power::TreeEvaluator::generateFillInDataBlockSequenceForUnresolvedField(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister)
+J9::Power::TreeEvaluator::generateTestAndReportFieldWatchInstructions(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *dataSnippet, bool isWrite, TR::Register *sideEffectRegister, TR::Register *valueReg, TR::Register *dataSnippetRegister)
    {
-   TR_ASSERT_FATAL(false, "This helper implements platform specific code for Fieldwatch, which is currently not supported on Power platforms.\n");
+   bool is64Bit = TR::Compiler->target.is64Bit();
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   bool isInstanceField = node->getOpCode().isIndirect();
+   bool isAOTCompile = cg->comp()->compileRelocatableCode();
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+
+   TR::Register *scratchReg = cg->allocateRegister();
+
+   // Check if snippet has been loaded
+   if (isInstanceField && !static_cast<TR::J9PPCWatchedInstanceFieldSnippet *>(dataSnippet)->isSnippetLoaded() ||
+      (!isInstanceField && !static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->isSnippetLoaded()))
+      loadFieldWatchSnippet(cg, node, dataSnippet, dataSnippetRegister, scratchReg, isInstanceField);
+
+   TR::Register *fieldClassReg = cg->allocateRegister();
+   TR::MemoryReference *classFlagsMemRef = NULL;
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+
+   TR::LabelSymbol* startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* fieldReportLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   uint8_t numOfConditions = getNumOfConditionsForReportFieldAccess(node, isResolved, isWrite,isInstanceField, cg);
+   numOfConditions++;
```

Comment:
What is the reason for this increment? Correct me if I'm wrong but this should be with the rest of the `getNumOfConditionsForReportFieldAccess` code right?

Reply:
@AlenBadel Why was this resolved? We can still make an improvement here. Instead of doing an unconditional increment here, we should instead incorporate this in the initialization above. Ex:
```
 // We need a minimum of two dependencies: data block and side effect register
uint8_t numOfConditions = 2;
```